docs/docs/typelevel.md

```diff
@@ -0,0 +1,532 @@
+   usual functional programming style in Scala.
+ - The ways to control implicit search are underdeveloped,
+   leading to complicated schemes, requiring rule prioritization or other hacks.
+ - Because of their conceptual complexity the resulting typelevel programs are often
+   fragile.
+ - Error diagnostics are usually very bad, making programming with implicits somewhat
+   of a black art. Dotty has greatly improved error dignostics for recursive implicits,
+   but the fundamental problem remains.
+
+In the following we develop a system and methodology that lets us do
+typelevel programming in a functional style. The system is based on
+Dotty's symmetric meta programming system, adding some fairly modest
+language extensions.
+
+### Prequel: Conditions in Staged Code
+
+As a perparatory step, we allow if-then-else in spliced code. Given an
+if-then-else expression, the macro-interpreter interprets its
+condition (which must be a compile-time constant) and chooses one of
+the two branches depending on the condition's value. Consider the
+following example mapping an integer value `n` to a peano number
+represented as an instance of type `Nat`.
+
+    trait Nat
+
+    case object Z extends Nat
+    case class S[N <: Nat] extends Nat
+
+    inline def toNat(inline n: Int): Nat = ~{
+      if n == 0 then 'Z
+      else 'S[toNat(n - 1)]
+    }
+
+In order to avoid clutter we use quote syntax without mandatory
+parentheses, e.g. `'Z` instead of `'(Z)`. (This assumes that symbol
+literals are confined to Scala-2 mode.)
+
+Note that the use of staging for `toNat` is recommended but not required. We could have
+written
+
+    inline def toNat(inline n: Int): Nat =
+      if n == 0 then Z
+      else S[toNat(n - 1)]
+
+But then we would rely on inlining's nornalizations to simplify the
+code to the branch that was taken. For more complicated examples this
+would not always work, so we'd end up sometimes with code being
+evaluated at runtime to compute a result that is already known at
+compile-time. Since typelevel programming is based on information
+known at compile-time, staging is a natural complement to it since it
+distinguishes syntactically between computations done at compile time
+and computations done at runtime. The syntactic distinctions also help
+understanding the code better because they make clear what gets
+computed when.
+
+The ability to use conditionals in spliced code does not add anything
+fundamental, since the same effect can be achieved by defining an
+intermediate function that contains the conditional and is called in a
+splice. By contrast, the next section does introduce a fundamental new
+capability.
+
+### First Step: Type Macros
+
+In the previous example we would like to establish a stronger link
+between the parameter `n` and the result type. For instance `toNat(3)`
+should have static type `S[S[S[Z]]]` instead of `Nat`. To
+get there, we allow macros that produce types. Just like an `inline
```

Comment:
Are these macros or type-level functions? And if the latter, is it F/FOmega functions or can you also pattern-match on types? Here you just show pattern-matching on compile-time values.

Reply:
They are macros that give rise to typelevel functions, as is explained in the following.